### PR TITLE
Vslt/vslt 74 start sc

### PR DESCRIPTION
**Pull request by xwa130saucer**
- [x] Able to start tests through ngrok
- [x] Able to start tests through sauce connect
- [x] Wrote tests for covering new added code

### CODE REVIEW

File: src/runner.js [hunk 1]
```diff
@@ -132,11 +132,14 @@ exports.run = function(config) {
     // establish ngrok tunnel and resolve ngrok url
```
Comment (loyalchow) [inline src/runner.js]: Change comment to `establish tunnel (ngrok or sauce connect)`

File: src/runner.js [hunk 2]
```diff
@@ -132,11 +132,14 @@ exports.run = function(config) {
...
     .then(function(proxyHost) {
       if (config.tunnel) {
-        console.log('Connecting tunnel');
-        return Tunnel.connect(proxyHost || config.tunnel.host, config.tunnel.token);
-      } else {
-        return Promise.resolve();
+        console.log('Connecting ngrok tunnel');
```
Comment (loyalchow) [inline src/runner.js]: Please change this back to just `Connecting tunnel`. Users may not know what "ngrok" is and get confused

File: src/runner.js [hunk 3]
```diff
@@ -132,11 +132,14 @@ exports.run = function(config) {
...
+        return Tunnel.connect({ ngrok: { host: proxyHost || config.tunnel.host, token: config.tunnel.token }});
+      }
+      if (config.sauce && config.sauce.launchSauceConnect) {
+        console.log('Connecting sauce tunnel');
```
Comment (loyalchow) [inline src/runner.js]: Update log to `Connecting Sauce Connect tunnel`

File: src/tunnel.js [hunk 4]
```diff
@@ -1,49 +1,79 @@
...
 var Promise = require('bluebird');
 var url = require('url');
+var sauceConnectLauncher = require('sauce-connect-launcher');
+var sauceConnection;
 
-exports.connect = function(host, token, tries = 0) {
-  if (!token) {
-    return Promise.reject(new Error('No Tunnel Token'));
+exports.connect = function({ ngrok, sauce }, tries = 0) {
+  if (sauce) {
```
Comment (eugenekatsov-sauce) [inline src/tunnel.js:8]: Hi, I was wondering about mixing function and fat arrows, should we settle on one approach going forward.

File: src/runner.js [hunk 5]
```diff
@@ -180,12 +183,16 @@ exports.run = function(config) {
...
       return api.waitForBuild(config.apiKey, config.project, config.branch, config.build).timeout(MAX_MS, 'Timeout waiting for Build');
     })
+    // disconnect tunnel and relay the response
     .then(function(response) {
-      clearInterval(timer);
-      if (config.tunnel) {
-        console.log('Disconnecting tunnel');
-        Tunnel.disconnect();
+      if (config.tunnel || (config.sauce && config.sauce.launchSauceConnect)) {
+        return Tunnel.disconnect(response);
```
Comment (loyalchow) [inline src/runner.js]: @xwa130saucer this `response` doesn't have anything to do with the tunnel, so it seems out of place. Can we not pass it in, and change it to something like the following instead?

```
return Tunnel.disconnect()
  .then(function() { return response; });
```

File: src/tunnel.js (hunk 4 above)
Comment (xwa130saucer) [reply src/tunnel.js:8]: Yeah, good point, it's kind of the code style guid, how do you think @loyalchow

File: src/runner.js [hunk 6]
```diff
@@ -180,12 +183,16 @@ exports.run = function(config) {
       timer = setInterval(function() { console.log('.'); }, 120*1000);
       return api.waitForBuild(config.apiKey, config.project, config.branch, config.build).timeout(MAX_MS, 'Timeout waiting for Build');
     })
+    // disconnect tunnel and relay the response
     .then(function(response) {
-      clearInterval(timer);
-      if (config.tunnel) {
-        console.log('Disconnecting tunnel');
```
Comment (loyalchow) [inline src/runner.js:186]: Let's keep this log

File: src/tunnel.js [hunk 7]
```diff
@@ -58,6 +88,16 @@ exports.transformUrl = function(origUrl, host, tunnelHost) {
   return newUrl;
 };
 
-exports.disconnect = function() {
-  ngrok.disconnect();
+exports.disconnect = function(response) {
+  return new Promise(function(resolve) {
+    if (sauceConnection) {
+      sauceConnection.close(function() {
+        console.log('Closed Sauce Connect process');
```
Comment (loyalchow) [inline src/tunnel.js]: Let's remove this log, and add back the general tunnel disconnecting log above

**Review by loyalchow — CHANGES_REQUESTED** [review]

File: src/tunnel.js (hunk 4 above)
Comment (loyalchow) [reply src/tunnel.js:8]: @xwa130saucer Let's use arrows going forward. But let's not update them all now - we can do so in a separate ticket. Also, we need to update the node requirements in package.json to node >= 6

Comment (xwa130saucer) [reply src/tunnel.js:8]: got it, and update the `engine >= 6`

File: src/runner.js (hunk 1 above)
Comment (xwa130saucer) [reply src/runner.js]: changed

File: src/runner.js (hunk 2 above)
Comment (xwa130saucer) [reply src/runner.js]: changed

File: src/runner.js (hunk 3 above)
Comment (xwa130saucer) [reply src/runner.js]: changed

File: src/runner.js (hunk 5 above)
Comment (xwa130saucer) [reply src/runner.js]: Good idea 👍 
changed

File: src/runner.js (hunk 6 above)
Comment (xwa130saucer) [reply src/runner.js:186]: added

File: src/tunnel.js (hunk 7 above)
Comment (xwa130saucer) [reply src/tunnel.js]: removed

File: src/runner.js [hunk 8]
```diff
@@ -129,14 +129,17 @@ exports.run = function(config) {
...
     .then(function(proxyHost) {
       if (config.tunnel) {
         console.log('Connecting tunnel');
-        return Tunnel.connect(proxyHost || config.tunnel.host, config.tunnel.token);
-      } else {
-        return Promise.resolve();
+        return Tunnel.connect({ ngrok: { host: proxyHost || config.tunnel.host, token: config.tunnel.token }});
+      }
+      if (config.sauce && config.sauce.launchSauceConnect) {
+        console.log('Connecting tunnel');
```
Comment (loyalchow) [inline src/runner.js]: Change to `Connecting Sauce Connect tunnel` to differentiate in the logs

**Review by loyalchow — APPROVED** [review]
One minor log change, and then can merge it

**Pull request merged**